vividus-plugin-email/src/main/java/org/vividus/bdd/email/service/ImapFetchService.java

```diff
@@ -0,0 +1,351 @@
+        {
+            return new PasswordAuthentication(username, password);
+        }
+    }
+
+    private final class PollingMessageListener implements MessageCountListener
+    {
+        private final Runnable refresh;
+        private final SearchTerm searchTerm;
+        private final List<Message> messages;
+
+        private final ReentrantLock lock = new ReentrantLock();
+        private final Condition messageArrivedCondition = lock.newCondition();
+        private final Condition messageHandleCondition = lock.newCondition();
+        private final AtomicBoolean handleFlag = new AtomicBoolean(true);
+
+        private PollingMessageListener(Runnable refresh, SearchTerm searchTerm)
+        {
+            this.refresh = refresh;
+            this.searchTerm = searchTerm;
+            this.messages = new CopyOnWriteArrayList<>();
+        }
+
+        @Override
+        public void messagesAdded(MessageCountEvent event)
+        {
+            try
+            {
+                lock.lock();
+                messageArrivedCondition.signal();
+                interruptible(() ->
+                {
+                    while (handleFlag.get())
+                    {
+                        messageHandleCondition.await();
+                    }
+                    return null;
+                });
+                for (Message message : event.getMessages())
+                {
+                    if (searchTerm.match(message))
+                    {
+                        messages.add(message);
+                    }
+                }
+            }
+            finally
+            {
+                try
+                {
+                    handleFlag.set(true);
+                    messageHandleCondition.signal();
+                }
+                finally
+                {
+                    lock.unlock();
+                }
+            }
+        }
+
+        @Override
+        public void messagesRemoved(MessageCountEvent event)
+        {
+            // empty
+        }
+
+        public List<Message> getMessages() throws InterruptedException
+        {
+            while (lock.isLocked())
+            {
+                Sleeper.sleep(Duration.ofSeconds(1));
+            }
+            try
+            {
+                lock.lock();
+                refresh.run();
+                boolean arrived = false;
+                while (true)
+                {
+                    arrived = messageArrivedCondition.await(messageEventWaitInMillis, TimeUnit.MILLISECONDS);
+                    break;
+                }
+
+                if (arrived)
+                {
+                    handleFlag.set(false);
+                    messageHandleCondition.signal();
+                    while (!handleFlag.get())
+                    {
+                        messageHandleCondition.await();
+                    }
+                }
+                return StreamSupport.stream(this.messages.spliterator(), false)
```

Comment:
new ArrayLst<>(this.messages) ?

Reply:
this.messages is the COWAL, just creating ArrayList from another collection doesn't guarantee thread safety